pkg/ccl/storageccl/add_sstable.go

```diff
@@ -61,39 +59,32 @@ func evalAddSSTable(
 	}
 	ms.Subtract(existingStats)
 
+	// To verify every KV is a valid roachpb.KeyValue in the range [start, end)
+	// we a) pass a verify flag on the iterator so that as ComputeStatsGo calls
+	// Next, we're also verifying each KV pair. We explicitly check the first key
+	// is >= start and then that we do not find a key after end.
+	dataIter, err := engineccl.NewMemSSTIterator(args.Data, true)
+	if err != nil {
+		return result.Result{}, err
+	}
+	defer dataIter.Close()
+
 	// Verify that the keys in the sstable are within the range specified by the
 	// request header, verify the key-value checksums, and compute the new
 	// MVCCStats.
 	stats, err := verifySSTable(
-		existingIter, args.Data, mvccStartKey, mvccEndKey, h.Timestamp.WallTime)
+		existingIter, dataIter, mvccStartKey, mvccEndKey, h.Timestamp.WallTime)
 	if err != nil {
 		return result.Result{}, errors.Wrap(err, "verifying sstable data")
 	}
 	ms.Add(stats)
 
-	return result.Result{
-		Replicated: storagebase.ReplicatedEvalResult{
-			AddSSTable: &storagebase.ReplicatedEvalResult_AddSSTable{
-				Data:  args.Data,
-				CRC32: util.CRC32(args.Data),
-			},
-		},
-	}, nil
+	return batcheval.MakeAddSSTableResult(ctx, args.Data, dataIter.Reader())
 }
 
 func verifySSTable(
-	existingIter engine.SimpleIterator, data []byte, start, end engine.MVCCKey, nowNanos int64,
+	existingIter, dataIter engine.SimpleIterator, start, end engine.MVCCKey, nowNanos int64,
```

Comment:
I'm a little wary that `verifySSTable` can still guarantee that it does what it is supposed to now, since it is now relying on dataIter being a verifying iter. Random thought: make it take the concrete type and inspect, or even toggle, the verify flag itself.

Reply:
alternatively, maybe we don't care about Verify since we made checksums optional?